File: src/php/PhantasyIslandPhpRemoteControl/radio/WaitToken.php
<?php

namespace PhantasyIslandPhpRemoteControl\Radio;

/**
 * 轻量级的请求-响应等待令牌的 PHP 实现。
 * PHP 不原生支持多线程等待，此处的 wait 采用轮询模拟。
 */
class WaitToken
{
    public $wait_cmd;
    public $time_base_id;
    public $response = null;
    public $processed_response = null;
    
    private $is_done = false;
    private $post_processor = null;

    public function __construct($wait_cmd, $time_base_id)
    {
        $this->wait_cmd = $wait_cmd;
        $this->time_base_id = $time_base_id;
    }

    public function isDone()
    {
        return $this->is_done;
    }

    public function setPostProcessor(callable $processor)
    {
        $this->post_processor = $processor;
        return $this;
    }

    private function applyPostProcessor($data)
    {
        if ($this->post_processor !== null) {
            $result = call_user_func($this->post_processor, $data);
        } else {
            $result = $data;
        }
        $this->processed_response = $result;
        return $result;
    }

    /**
     * 由 RadioManager 在收到消息时调用。
     */
    public function complete($data)
    {
        $this->response = $data;
        $this->applyPostProcessor($data);
        $this->is_done = true;
    }

    /**
     * 阻塞当前进程直到收到响应或超时（轮询实现）。
     * @param float $timeout 超时秒数
     * @return mixed|null 后处理结果，超时返回 null
     */
    public function wait($timeout = 3.0)
    {
        $start = microtime(true);
        while (!$this->is_done) {
            if (microtime(true) - $start > $timeout) {
                return null;
            }
            // 在实际使用中，这里需要某种方式让 Socket.IO 客户端有机会处理入站消息
            // 例如：$this->rm->socket->work();
            usleep(10000); // 10ms
        }
        return $this->processed_response;
    }
}

File: src/el/radio/wait_token.erl
-module(wait_token).
-behaviour(gen_server).

-export([start_link/2, complete/2, wait/2, set_post_processor/2]).
-export([init/1, handle_call/3, handle_cast/2]).

-record(state, {
    wait_cmd,
    time_base_id,
    response = nil,
    processed_response = nil,
    post_processor = nil,
    waiting_from = nil
}).

start_link(WaitCmd, TimeBaseID) ->
    gen_server:start_link(?MODULE, [WaitCmd, TimeBaseID], []).

complete(Pid, Data) ->
    gen_server:cast(Pid, {complete, Data}).

wait(Pid, Timeout) ->
    try
        gen_server:call(Pid, wait, Timeout)
    catch
        exit:{timeout, _} -> nil
    end.

set_post_processor(Pid, Processor) ->
    gen_server:cast(Pid, {set_post_processor, Processor}).

init([WaitCmd, TimeBaseID]) ->
    {ok, #state{wait_cmd = WaitCmd, time_base_id = TimeBaseID}}.

handle_cast({set_post_processor, Processor}, State) ->
    {noreply, State#state{post_processor = Processor}};

handle_cast({complete, Data}, State = #state{post_processor = PP, waiting_from = From}) ->
    Processed = if PP =/= nil -> PP(Data); true -> Data end,
    if From =/= nil -> gen_server:reply(From, Processed); true -> ok end,
    {noreply, State#state{response = Data, processed_response = Processed, waiting_from = nil}}.

handle_call(wait, From, State = #state{response = Resp, processed_response = ProcResp}) ->
    case Resp of
        nil -> {noreply, State#state{waiting_from = From}};
        _ -> {reply, ProcResp, State}
    end.
